### PR TITLE
DRILL-4782 / DRILL-7139: Fix cast and arithmetic for DayInterval 

### PR DESCRIPTION
**Pull request by dgrinchenko**
- cast function for the day interval changed to round milliseconds to complete days
- ToDateTypeFunctions#toTime now returning milliseconds of day
- updated the way how DayInterval subtracts and adds, to follow the cast function logic

UT core updates: 
- added vectorValue function to the queryBuilder to simplify retrieving value of the vector
- refactored singleton query result functions at queryBuilder

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@ihuzenko please review.

File: exec/java-exec/src/main/codegen/templates/CastStringTypesToInterval.java [hunk 1]
```diff
@@ -86,24 +86,35 @@ public void eval() {
...
-                       (period.getMinutes() * org.apache.drill.exec.vector.DateUtilities.minutesToMillis) +
-                       (period.getSeconds() * org.apache.drill.exec.vector.DateUtilities.secondsToMillis) +
-                       (period.getMillis());
+    long millis = period.getHours() * (long) org.apache.drill.exec.vector.DateUtilities.hoursToMillis +
+                  period.getMinutes() * (long) org.apache.drill.exec.vector.DateUtilities.minutesToMillis +
+                  period.getSeconds() * (long) org.apache.drill.exec.vector.DateUtilities.secondsToMillis +
+                  period.getMillis();
+
+    // Subtracts all full days from millis
+    if (millis > org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis - 1) {
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/CastStringTypesToInterval.java]:
```suggestion
    if (millis >= org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis) {
      // Extract days from millis and add them to out.days
      int daysFromMillis = (int) (millis / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
      millis -= daysFromMillis * org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
      out.days += daysFromMillis;
    }
```

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 2]
```diff
@@ -64,8 +64,17 @@ public void eval() {
             <#elseif intervaltype == "IntervalYear">
             out.value = left.value + right.value;
             <#elseif intervaltype == "IntervalDay">
+            long millis = (long) left.milliseconds + right.milliseconds;
             out.days = left.days + right.days;
-            out.milliseconds = left.milliseconds + right.milliseconds;
+
+            // Subtracts all full days from millis
+            if (millis > org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis - 1) {
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]:
```suggestion
   if (millis >= org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis) {
     // Extract days from millis and add them to out.days
     int daysFromMillis = (int) (millis / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis);
     millis -= daysFromMillis * org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
     out.days += daysFromMillis;
   }
```

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 3]
```diff
@@ -64,8 +64,17 @@ public void eval() {
             <#elseif intervaltype == "IntervalYear">
             out.value = left.value + right.value;
             <#elseif intervaltype == "IntervalDay">
+            long millis = (long) left.milliseconds + right.milliseconds;
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]: Please move the line down to be closer to if statement

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 4]
```diff
@@ -64,8 +64,17 @@ public void eval() {
...
             out.days = left.days + right.days;
-            out.milliseconds = left.milliseconds + right.milliseconds;
+
+            // Subtracts all full days from millis
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]:
```suggestion
           
```

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 5]
```diff
@@ -90,6 +99,21 @@ public void eval() {
             <#elseif intervaltype == "IntervalDay">
             out.days = left.days - right.days;
             out.milliseconds = left.milliseconds - right.milliseconds;
+
+            // redistribute days stored in millis to days
+            int daysFromMillis = Math.abs(out.milliseconds) / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]:
```suggestion
    int daysFromMillis = out.milliseconds / org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis;
    if (daysFromMillis != 0) {
      out.milliseconds -= org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis * daysFromMillis;
      out.days -= Math.abs(daysFromMillis);
    }
```

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 6]
```diff
@@ -90,6 +99,21 @@ public void eval() {
...
+            if (out.milliseconds > 0) {
+              out.milliseconds = out.milliseconds - (org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis * daysFromMillis);
+            } else {
+              out.milliseconds = out.milliseconds + (org.apache.drill.exec.vector.DateUtilities.daysToStandardMillis * daysFromMillis);
+            }
+            out.days -= daysFromMillis;
+
+            // if milliseconds are bellow zero, redistribute them in count of whole day
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]:
```suggestion
           // if milliseconds are bellow zero, try to subtract them from days
```

File: exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java [hunk 7]
```diff
@@ -90,6 +99,21 @@ public void eval() {
...
+            if (out.days > 0 && out.milliseconds < 0) {
```
Comment (ihuzenko) [inline exec/java-exec/src/main/codegen/templates/DateIntervalFunctionTemplates/IntervalIntervalArithmetic.java]:
```suggestion
            if (out.milliseconds < 0 && out.days > 0) {
```

**Pull request closed without merging**